### PR TITLE
[Block] Add option to configure block device flush

### PR DESCRIPTION
**Pull request by georgepisaltu**
# Reason for This PR

#2172 

## Description of Changes

This commit introduces the notion of cache type to the virtio block device implementation. This allows users to choose from two cache types:

- `unsafe`: this option will advertise the flush virtio capability to the guest block driver, but will acknowledge and ignore incoming flush requests;
- `writethrough`: this option will advertise the flush virtio capability to the guest block driver and will perform a `fsync` syscall on incoming flush requests;

This functionality is exposed to users through an optional `ignore_flush` field in the block configuration API request. When absent, this field defaults to `true`, ignoring incoming flush requests from the guest driver and keeping the behavior of the block device the same as before this change. Setting it to `false` will make the device call `fsync` on the backing file when processing flush requests. 

- [x] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

### CODE REVIEW

File: src/devices/src/virtio/block/device.rs
Comment (acatangiu) [inline src/devices/src/virtio/block/device.rs:42]: `writethrough` is not the best name since that usually means **no host FS caching** which is not the case here.

Unfortunately I don't have a better name right now 😄 Suggestions are welcome!

File: src/devices/src/virtio/block/device.rs
Comment (acatangiu) [inline src/devices/src/virtio/block/device.rs:177]: Why not use the same `enum CacheType` here as well? Ideally even use it from API all the way down without transformations.

File: src/devices/src/virtio/block/persist.rs
Comment (acatangiu) [inline src/devices/src/virtio/block/persist.rs:44]: Should we make this backward-compatible, where the cache/flush just reverts to old behavior (also print a `warn!` or smth)?

File: src/api_server/swagger/firecracker.yaml
Comment (acatangiu) [inline src/api_server/swagger/firecracker.yaml:760]: Why not use the `CacheType` enum here instead of this very specific bool?

The enum is forward-looking as it allows us to add options to it or change implementation details for each `CacheType`.
This bool is clearer as it is self-explanatory, but also limited and not extensible.

File: src/devices/src/virtio/block/device.rs
Comment (alindima) [reply src/devices/src/virtio/block/device.rs:42]: I think the right naming is `writeback`. This is the term QEMU uses as well: https://documentation.suse.com/sles/11-SP4/html/SLES-kvm4zseries/cha-qemu-cachemodes.html#id-1.10.6.6.5

Comment (georgepisaltu) [reply src/devices/src/virtio/block/device.rs:42]: `writeback` is indeed the right name. I copy-pasted the wrong one the first time around 😆. Nice catch.

File: src/devices/src/virtio/block/device.rs
Comment (georgepisaltu) [reply src/devices/src/virtio/block/device.rs:177]: Only reason was that `ignore_flush` seemed like a perfect name for the field.

Changed it so that we use `cache_type` as a field in the API request JSON, all the way down to the device.

File: src/devices/src/virtio/block/persist.rs
Comment (georgepisaltu) [reply src/devices/src/virtio/block/persist.rs:44]: I am leaning towards no because of the risk that users might shoot themselves in the foot, but I can see how we can enable this and print a warning or something and let them handle it.

I guess we can do either, I'm on the fence about this. Other opinions?

File: src/api_server/swagger/firecracker.yaml
Comment (georgepisaltu) [reply src/api_server/swagger/firecracker.yaml:760]: Changed it so that it uses `cache_type`.

File: src/devices/src/virtio/block/persist.rs
Comment (acatangiu) [reply src/devices/src/virtio/block/persist.rs]: @firecracker-microvm/compute-capsule

Comment (georgepisaltu) [reply src/devices/src/virtio/block/persist.rs]: I went ahead and changed it so that it's only a warning now. Also added an integ test for this case.

@firecracker-microvm/compute-capsule please raise concerns if you have any.

**Comment by alindima** [comment]
This PR also partially fixes: #2207.
In order to fully fix it, we should also tackle in this PR this related issue: https://github.com/firecracker-microvm/firecracker/issues/2188

The block device caching strategy could also influence the decision of calling `fsync` on `DiskProperties::drop`, so that we make sure that all data has been flushed before closing the backing file, WDYT @acatangiu @georgepisaltu ?

**Comment by acatangiu** [comment]
> This PR also partially fixes: #2207.
> In order to fully fix it, we should also tackle in this PR this related issue: #2188
> 
> The block device caching strategy could also influence the decision of calling `fsync` on `DiskProperties::drop`, so that we make sure that all data has been flushed before closing the backing file, WDYT @acatangiu @georgepisaltu ?

@2207 is fully fixed by this PR in my view. We have seen no customer usecase for `O_SYNC/O_DSYNC` so we can update the issue and close it as fixed by this PR.

Indeed, for completeness, we should add the `fsync` on `DiskProperties::drop` for disks where configured caching is `writeback`.

**Comment by alindima** [comment]
> > This PR also partially fixes: #2207.
> > In order to fully fix it, we should also tackle in this PR this related issue: #2188
> > The block device caching strategy could also influence the decision of calling `fsync` on `DiskProperties::drop`, so that we make sure that all data has been flushed before closing the backing file, WDYT @acatangiu @georgepisaltu ?
> 
> @2207 is fully fixed by this PR in my view. We have seen no customer usecase for `O_SYNC/O_DSYNC` so we can update the issue and close it as fixed by this PR.
> 
> Indeed, for completeness, we should add the `fsync` on `DiskProperties::drop` for disks where configured caching is `writeback`.

Yes, I agree. I meant that we can close #2207 if we also add the `drop` implementation, not necessarily `O_SYNC` support

**Comment by georgepisaltu** [comment]
Added a `fsync` call on `DiskProperties::drop`.

File: src/vmm/src/version_map.rs
Comment (acatangiu) [inline src/vmm/src/version_map.rs:21]: I think this commit should be squashed together with the one changing block state struct.

File: src/devices/src/virtio/block/persist.rs
Comment (acatangiu) [inline src/devices/src/virtio/block/persist.rs:28]: nitish: this being a simple enum, we could have just used the original instead of creating a shadow _state_ one.

File: src/devices/src/virtio/block/persist.rs
Comment (acatangiu) [inline src/devices/src/virtio/block/persist.rs:43]: nit: idiomatically better to implement [std::convert::From](https://doc.rust-lang.org/std/convert/trait.From.html).

File: src/devices/src/virtio/block/persist.rs
Comment (acatangiu) [inline src/devices/src/virtio/block/persist.rs:54]: do we need a default_fn ?

File: src/devices/src/virtio/block/persist.rs
Comment (acatangiu) [inline src/devices/src/virtio/block/persist.rs:93]: nit: could have been
```diff
- cache_type: CacheTypeState::from_cache_type(self.cache_type()),
+ cache_type: self.cache_type().into(),
```

File: tests/integration_tests/build/test_coverage.py
Comment (acatangiu) [inline tests/integration_tests/build/test_coverage.py:26]: pls coverage commit to integ tests commit (or some other)

File: tests/integration_tests/performance/test_versioned_serialization_benchmark.py
Comment (acatangiu) [inline tests/integration_tests/performance/test_versioned_serialization_benchmark.py:31]: pls explain in commit why we're doing a 25% increase in delta here.

For sanity, please run the benchmark (can even do it locally) with and without these changes and document in commit description if/what are the changes' impact if any.

**Review by acatangiu — COMMENTED** [review]
Code looks good, only nits.

PR is missing documentation. The newly available caching strategies need to be included in block device documentation.

File: src/devices/src/virtio/block/persist.rs
Comment (acatangiu) [reply src/devices/src/virtio/block/persist.rs:54]: yes, we need it, nvm

File: src/devices/src/virtio/block/persist.rs
Comment (georgepisaltu) [reply src/devices/src/virtio/block/persist.rs:28]: My understanding is that we should not include any `persist` logic in our devices, so it didn't make sense to add a `#[derive(Versionize)]` on `CacheType`.

Furthermore, `CacheType` is a part of the block device and has little to do with any snapshot logic, so it didn't make sense to have it in `block/persist.rs` either.

This was the reasoning behind creating a shadow struct for `CacheType`.

File: src/devices/src/virtio/block/persist.rs
Comment (georgepisaltu) [reply src/devices/src/virtio/block/persist.rs:43]: Done.

File: src/devices/src/virtio/block/persist.rs
Comment (georgepisaltu) [reply src/devices/src/virtio/block/persist.rs:93]: Done.

File: src/vmm/src/version_map.rs
Comment (georgepisaltu) [reply src/vmm/src/version_map.rs:21]: Done.

File: tests/integration_tests/build/test_coverage.py
Comment (georgepisaltu) [reply tests/integration_tests/build/test_coverage.py:26]: Done.

File: src/devices/src/virtio/block/persist.rs
Comment (sandreim) [inline src/devices/src/virtio/block/persist.rs:52]: This is actually version 2 as it refers to the struct version, not the app version.

File: src/devices/src/virtio/block/persist.rs
Comment (sandreim) [inline src/devices/src/virtio/block/persist.rs:65]: This needs coverage in unit test.

File: src/vmm/src/version_map.rs
Comment (sandreim) [reply src/vmm/src/version_map.rs:21]: BlockState has version 2.

File: src/devices/src/virtio/block/persist.rs
Comment (sandreim) [inline src/devices/src/virtio/block/persist.rs:76]: This also needs coverage.

File: tests/integration_tests/performance/test_versioned_serialization_benchmark.py
Comment (georgepisaltu) [reply tests/integration_tests/performance/test_versioned_serialization_benchmark.py:31]: I ran it and in 23 tries there was one failure on serializing with CRC on Intel `x86_64`. The test was off by `0.00001`, or about 0.05% of the original delta.

Nevertheless, the fail happened and this change is to ensure we don't have any intermittent failures on this in the future.

While adding a bit of extra logic in `block/persist.rs` certainly won't speed ser/deser up, it doesn't seem to degrade the averages too much. From my measurements, there seems to be a difference of about `0.0001` on average between the means before and after the change, which is less than 0.1%. I believe this failure could have happened without the change simply because the bench parameters were a bit too strict, or it may have just been a one time thing where the machine I was testing on had a lazy CPU cycle or two :grinning: .

**Review by georgepisaltu — COMMENTED** [review]
Also added the documentation in `docs/api_requests/` as the block device doesn't have its own doc. I will create the documentation for the block device and move the cache documentation in a subsequent PR to not block this one as they are not related.

File: src/devices/src/virtio/block/persist.rs
Comment (georgepisaltu) [reply src/devices/src/virtio/block/persist.rs:54]: Changed it.

File: src/devices/src/virtio/block/persist.rs
Comment (georgepisaltu) [reply src/devices/src/virtio/block/persist.rs:67]: It already has coverage in a unit test.

File: src/devices/src/virtio/block/persist.rs
Comment (georgepisaltu) [reply src/devices/src/virtio/block/persist.rs:78]: Added a unit test for this.

File: src/vmm/src/version_map.rs
Comment (georgepisaltu) [reply src/vmm/src/version_map.rs:21]: Changed it.

File: tests/integration_tests/functional/test_drives.py
Comment (alindima) [inline tests/integration_tests/functional/test_drives.py:370]: actually, the commands are generated but they are ignored

File: src/devices/src/virtio/block/device.rs
Comment (acatangiu) [inline src/devices/src/virtio/block/device.rs:200]: this is already part of `DiskProperties`, why duplicate here?

File: src/devices/src/virtio/block/persist.rs
Comment (acatangiu) [reply src/devices/src/virtio/block/persist.rs:28]: > My understanding is that we should not include any persist logic in our devices

I don't think that's accurate. Plain-old-data structs can be versioned directly I think. @sandreim can confirm/disprove this, I might be missing something.

nit: In current PR `CacheType` already has versionize.

File: docs/api_requests/block-caching.md
Comment (alindima) [inline docs/api_requests/block-caching.md:23]: Maybe note here that the data may is only committed to the physical disk when the host driver decides to

File: src/devices/src/virtio/block/persist.rs
Comment (sandreim) [reply src/devices/src/virtio/block/persist.rs:28]: My take on this is to keep our devices clean of any serialization logic. This is mostly forward looking towards ingesting upstream rust-vmm devices more easily.

File: docs/api_requests/block-caching.md
Comment (georgepisaltu) [reply docs/api_requests/block-caching.md:23]: I think an explanation of how caching works on Linux systems is not appropriate here, especially since whatever we write in this section highly depends on how the host OS and FS are set up. The fact that IO is cached is the default behavior in Linux systems and I don't think we need to explain this here.

We should, however, touch on this in the general block device documentation and provide our recommendations regarding host OS and FS setup. In that section we should also mention what host configuration we are running our tests on, and that should be enough.

File: src/api_server/src/request/drive.rs
Comment (georgepisaltu) [reply src/api_server/src/request/drive.rs:245]: This is how `serde::Serialize` works. The enum variants are `Unsafe` and `Writeback` due to Rust style and `serde` just picks this up and converts it to a string JSON.

There is also precedence regarding upper-case JSON fields when using `serde`, the most common example being `snapshot_type: "Full"`.

File: src/devices/src/virtio/block/device.rs
Comment (georgepisaltu) [reply src/devices/src/virtio/block/device.rs:171]: There is no need to include `file.flush()` because `std::fs::File` is not buffered and that is by design. It works the same way an FD would work in C when using `open`, `read` and `write`.

For buffered IO in C, one would use `fopen`, `fread` and `fwrite`. In Rust, this is achieved by using `BufReader` and `BufWriter`.

This means that all data written with a `file.write()` will reach the cache as long as the file is not opened with `O_DIRECT` or `O_DSYNC` (which is not the case in Firecracker), and this is exactly the reason the Rust `file.flush()` is a noop. You can check this and the non-buffered behavior of `file.flush()` using `strace`.

Moreover, I don't think the argument that the `File` implementation might change supports adding `file.flush()`. If anything, it prohibits it. Right now, `file.flush()` is a noop, which means the behviour is the same as not having it. If it does _anything_ in the future and we add it now in the implementation, it would create another bug.

File: src/devices/src/virtio/block/device.rs
Comment (georgepisaltu) [reply src/devices/src/virtio/block/device.rs:200]: You're right, I removed the duplicate.

File: src/devices/src/virtio/block/device.rs
Comment (georgepisaltu) [reply src/devices/src/virtio/block/device.rs:440]: You're right, I removed the duplicate.

File: src/vmm/src/builder.rs
Comment (georgepisaltu) [reply src/vmm/src/builder.rs:1300]: Should have been `Writeback` in the code, the comment was right. Changed it.

File: src/vmm/src/rpc_interface.rs
Comment (georgepisaltu) [reply src/vmm/src/rpc_interface.rs:1070]: Because the purpose of this test is to ensure a request to add a second root device will fail, it does not test any block capability.

The parsing side of this is tested in the `api_server`.

File: src/vmm/src/vmm_config/drive.rs
Comment (georgepisaltu) [reply src/vmm/src/vmm_config/drive.rs:96]: I thought we can't but it seems `serde` is smarter than I thought. Changed it.

File: tests/integration_tests/functional/test_drives.py
Comment (georgepisaltu) [reply tests/integration_tests/functional/test_drives.py:370]: Indeed, changed it.

File: src/api_server/src/request/drive.rs
Comment (alindima) [reply src/api_server/src/request/drive.rs:245]: I was thinking that having case-sensitive values is more error-prone. and serde has a rename directive that can take lowercase as value: for example one can add this on the enum:

`#[serde(rename_all = "lowercase")]`

I'm fine with keeping it CamelCase though.

File: src/vmm/src/version_map.rs
Comment (sandreim) [inline src/vmm/src/version_map.rs:31]: Please also include the binary artifacts for running snapshot integration tests across versions.

Comment (georgepisaltu) [reply src/vmm/src/version_map.rs:31]: Once I get all approves, I will build and publish the artifacts so they are built from the same source as what is going to get merged.

File: src/devices/src/virtio/block/device.rs
Comment (alindima) [reply src/devices/src/virtio/block/device.rs:171]: My arguments were also rooted in the fact that for `CacheType::WriteBack`, we do a `file.flush()` before calling f`sync()`. why is that?

Even if this a noop, this just follows the semantic of the `flush` method of the `Write` trait

Comment (georgepisaltu) [reply src/devices/src/virtio/block/device.rs:171]: It is just future proofing. I thought about the semantics of `File::flush` possibly changing in the future and I wanted to be on the safe side. The rationale was that it has no performance cost now, as it is a noop, and could save us some trouble in the future in the unlikely event that flush semantics change.

I would agree to removing it, I have no strong preference here. Let me know what you think.

**Review by acatangiu — APPROVED** [review]

**Review by alindima — APPROVED** [review]

**Pull request merged**
